integration_test/ESI/formats.py

```diff
@@ -0,0 +1,242 @@
+
+    def run(self):
+        """Run the test by creating a Python script, starting the simulation,
+        running the Python script, then stopping the simulation.
+
+        Not perfect since we don't know when the cosim RPC server in the
+        simulation has started accepting connections."""
+
+        # Find available port.
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.bind(('', 0))
+        port = sock.getsockname()[1]
+        sock.close()
+
+        with open(os.path.join(self.execdir, "script.py"), "w") as script:
+            # Include a bunch of config variables at the beginning of the
+            # script for use by the test code.
+            cfg = self.test.config
+            allDicts = list(self.test.config.__dict__.items()) + \
+                list(self.litConfig.__dict__.items())
+            vars = dict([(key, value)
+                         for (key, value) in allDicts if isinstance(value, str)])
+            vars["execdir"] = self.execdir
+            vars["srcdir"] = self.srcdir
+            vars["srcfile"] = self.file
+            # 'rpcSchemaPath' points to the CapnProto schema for RPC and is the
+            # one that nearly all scripts are going to need.
+            vars["rpcschemapath"] = os.path.join(
+                cfg.circt_include_dir, "circt", "Dialect", "ESI", "cosim",
+                "CosimDpi.capnp")
+            script.writelines(f"{name} = \"{value}\"\n" for (
+                name, value) in vars.items())
+            script.write("\n\n")
+
+            # Add the test files directory and this files directory to the
+            # pythonpath.
+            script.write(f"import os\n")
+            script.write(f"import sys\n")
+            script.write(
+                f"sys.path.append(\"{os.path.dirname(self.file)}\")\n")
+            script.write(
+                f"sys.path.append(\"{os.path.dirname(__file__)}\")\n")
+            script.write("\n\n")
+            script.write(
+                "simhostport = f'{os.uname()[1]}:" + str(port) + "'\n")
+
+            # Run the lines specified in the test file.
+            script.writelines(
+                f"{x}\n" for x in self.runs)
+
+        timedout = False
+        try:
+            # Run the simulation.
+            simStdout = open(os.path.join(self.execdir, "sim_stdout.log"), "w")
+            simStderr = open(os.path.join(self.execdir, "sim_stderr.log"), "w")
+            simEnv = os.environ.copy()
+            simEnv["COSIM_PORT"] = str(port)
+            simProc = subprocess.Popen(
+                [f"./obj_dir/V{self.top}", "--cycles", "-1"],
+                stdout=simStdout, stderr=simStderr, cwd=self.execdir,
+                env=simEnv)
+            # Wait a set amount of time for the simulation to start accepting
+            # RPC connections.
+            # TODO: Check if the server is up by polling.
+            time.sleep(0.05)
```

Comment:
These cosim integration tests won't be a part of the tests we run on GitHub for the PR gate right? Leaving something like this is probably fine if you (and probably me :)) are the only users, but if it could lead to non-deterministic tests on GitHub, I'd say let's tackle the TODO before merge.

Reply:
They are not currently run in the PR gates (none of the integration tests do) nor do we run nightly integration tests. Yeah, this'll be a problem if we scale further.

Reply:
I've added #240 to track this.